File: frontend/src/types/diagnostico.ts
export interface RespuestaCorrecta {
  codigo: string;
  respuesta_correcta: string;
}

export interface RespuestaCorrectaBatch {
  materia: string;
  periodo: string;
  respuestas: RespuestaCorrecta[];
}

export interface DiagnosticoAlumnoResultado {
  alumno_id: number;
  nombre_completo: string;
  numero_cuenta: string | null;
  numero_folio: string | null;
  correo: string | null;
  ingenieria: string | null;
  respuestas_algebra: Array<{ codigo: string; respuesta: string; correcta: boolean }> | null;
  respuestas_trigonometria: Array<{ codigo: string; respuesta: string; correcta: boolean }> | null;
  respuestas_geometria: Array<{ codigo: string; respuesta: string; correcta: boolean }> | null;
  respuestas_calculo: Array<{ codigo: string; respuesta: string; correcta: boolean }> | null;
  puntaje_algebra: number | null;
  puntaje_trigonometria: number | null;
  puntaje_geometria: number | null;
  puntaje_calculo: number | null;
  promedio_diagnostico: number | null;
}

export interface DiagnosticoNoEncontrado {
  nombre_original: string;
  correo: string | null;
  cuenta: string | null;
  folio: string | null;
  materia: string;
  motivo: string;
  candidatos: Array<Record<string, unknown>>;
  indice: number;
}

export interface ResultadoProcesamientoDiagnostico {
  materia: string;
  periodo: string;
  total_filas: number;
  encontrados: number;
  no_encontrados: number;
  resultados: DiagnosticoAlumnoResultado[];
  no_encontrados_detalle: DiagnosticoNoEncontrado[];
}

export interface CorregirMatchingPayload {
  correcciones: Array<{
    indice: number;
    alumno_id: number;
  }>;
}

File: app/schemas/diagnostico.py
from pydantic import BaseModel, ConfigDict, Field


class RespuestaCorrectaBase(BaseModel):
    materia: str = Field(min_length=1, max_length=20)
    codigo: str = Field(min_length=1, max_length=10)
    respuesta_correcta: str = Field(pattern=r"^[abcd]$")
    periodo: str = Field(min_length=1, max_length=10)


class RespuestaCorrectaCreate(RespuestaCorrectaBase):
    pass


class RespuestaCorrectaRead(RespuestaCorrectaBase):
    model_config = ConfigDict(from_attributes=True)
    id: int


class RespuestaCorrectaBatch(BaseModel):
    respuestas: list[RespuestaCorrectaCreate]


class RespuestaPorCodigo(BaseModel):
    codigo: str
    respuesta: str


class RespuestaAlumnoPorMateria(BaseModel):
    codigo: str
    respuesta: str
    correcta: bool


class DiagnosticoAlumnoResultado(BaseModel):
    alumno_id: int
    nombre_completo: str
    numero_cuenta: str | None
    numero_folio: str | None
    correo: str
    ingenieria: str
    respuestas_algebra: list[RespuestaAlumnoPorMateria] | None = None
    respuestas_trigonometria: list[RespuestaAlumnoPorMateria] | None = None
    respuestas_geometria: list[RespuestaAlumnoPorMateria] | None = None
    respuestas_calculo: list[RespuestaAlumnoPorMateria] | None = None
    puntaje_algebra: float | None = None
    puntaje_trigonometria: float | None = None
    puntaje_geometria: float | None = None
    puntaje_calculo: float | None = None
    promedio_diagnostico: float | None = None


class DiagnosticoNoEncontrado(BaseModel):
    nombre_original: str
    correo: str | None
    cuenta: str | None
    folio: str | None
    materia: str
    motivo: str
    candidatos: list[dict] = []


class ResultadoProcesamientoDiagnostico(BaseModel):
    materia: str
    periodo: str
    total_filas: int
    encontrados: int
    no_encontrados: int
    resultados: list[DiagnosticoAlumnoResultado]
    no_encontrados_detalle: list[DiagnosticoNoEncontrado]


class CorregirMatchingItem(BaseModel):
    indice: int
    alumno_id: int


class CorregirMatchingPayload(BaseModel):
    materia: str
    periodo: str
    correcciones: list[CorregirMatchingItem]


class BuscarAlumnoResult(BaseModel):
    id: int
    nombre: str
    apellido_paterno: str
    apellido_materno: str
    correo: str | None
    numero_cuenta: str | None
    numero_folio: str | None
    ingenieria: str | None


class CrearAlumnoDiagnostico(BaseModel):
    nombre: str = Field(min_length=1, max_length=100)
    apellido_paterno: str = Field(min_length=1, max_length=100)
    apellido_materno: str = Field(min_length=1, max_length=100)
    correo_personal: str = Field(min_length=5, max_length=200)
    numero_cuenta: str | None = Field(default=None, max_length=7)
    numero_folio: str | None = Field(default=None, max_length=9)
    ingenieria_clave: str | None = Field(default=None, max_length=10)
    periodo: str = Field(min_length=1, max_length=10)
